File: secure_env.py
"""Secure .env storage using keyring (Windows Credential Manager / macOS Keychain)."""

import base64
import os
import sys

try:
    import keyring
except ImportError:
    keyring = None

SERVICE = 'AnthropicRichPresence'
ACCOUNT = 'env'
ENV_PATH = os.path.join(os.path.dirname(os.path.abspath(__file__)), '.env')


def encrypt():
    if not os.path.exists(ENV_PATH):
        print('.env file not found')
        sys.exit(1)
    if not keyring:
        print('keyring not installed: pip install keyring')
        sys.exit(1)
    with open(ENV_PATH, 'r', encoding='utf-8') as f:
        content = f.read()
    b64 = base64.b64encode(content.encode('utf-8')).decode('ascii')
    keyring.set_password(SERVICE, ACCOUNT, b64)
    user = os.environ.get('USER') or os.environ.get('USERNAME', 'unknown')
    print(f'Encrypted .env -> keyring ({SERVICE}, tied to {user})')
    print('You can now delete .env if desired.')


def decrypt():
    if not keyring:
        return None
    try:
        b64 = keyring.get_password(SERVICE, ACCOUNT)
        if not b64:
            return None
        return base64.b64decode(b64).decode('utf-8')
    except Exception as e:
        print(f'Failed to decrypt from keyring: {e}', file=sys.stderr)
        return None


def load_secure_env():
    content = decrypt()
    if not content:
        return False
    for line in content.splitlines():
        line = line.strip()
        if not line or line.startswith('#'):
            continue
        eq = line.find('=')
        if eq < 0:
            continue
        key = line[:eq].strip()
        value = line[eq + 1:].strip()
        if key not in os.environ:
            os.environ[key] = value
    return True


if __name__ == '__main__':
    cmd = sys.argv[1] if len(sys.argv) > 1 else ''
    if cmd == 'encrypt':
        encrypt()
    elif cmd == 'decrypt':
        content = decrypt()
        if content:
            print('[WARNING] Sensitive data follows:', file=sys.stderr)
            print(content)
        else:
            print('No credentials found or decryption failed')
    else:
        print('Usage: python secure_env.py <encrypt|decrypt>')

File: secure-env.js
// Secure .env storage - DPAPI on Windows, Keychain on macOS
// Encrypts .env values tied to the current user account

const fs = require('fs');
const path = require('path');
const { execFileSync } = require('child_process');

const IS_MACOS = process.platform === 'darwin';
const ENV_PATH = path.join(__dirname, '.env');
const SECURE_PATH = path.join(__dirname, '.env.encrypted');
const KEYCHAIN_SERVICE = 'AnthropicRichPresence';
const KEYCHAIN_ACCOUNT = 'env';

// --- Platform-specific encrypt/decrypt (no shell interpolation) ---

function platformEncrypt(plaintext) {
  const b64 = Buffer.from(plaintext, 'utf8').toString('base64');

  if (IS_MACOS) {
    try { execFileSync('security', ['delete-generic-password', '-s', KEYCHAIN_SERVICE, '-a', KEYCHAIN_ACCOUNT], { encoding: 'utf8', stdio: 'ignore' }); } catch {}
    execFileSync('security', ['add-generic-password', '-s', KEYCHAIN_SERVICE, '-a', KEYCHAIN_ACCOUNT, '-w', b64], { encoding: 'utf8' });
    return 'keychain';
  } else {
    const ps = `$ss = ConvertTo-SecureString -String '${b64}' -AsPlainText -Force; ConvertFrom-SecureString $ss`;
    return execFileSync('powershell', ['-NoProfile', '-Command', ps], { encoding: 'utf8' }).trim();
  }
}

function platformDecrypt(encrypted) {
  if (IS_MACOS || encrypted === 'keychain') {
    const b64 = execFileSync('security', ['find-generic-password', '-s', KEYCHAIN_SERVICE, '-a', KEYCHAIN_ACCOUNT, '-w'], { encoding: 'utf8' }).trim();
    return Buffer.from(b64, 'base64').toString('utf8');
  } else {
    const cleaned = encrypted.replace(/\s/g, '');
    if (!/^[0-9a-fA-F]+$/.test(cleaned)) {
      throw new Error('Invalid encrypted data format');
    }
    const ps = `$ss = ConvertTo-SecureString -String '${cleaned}'; $ptr = [System.Runtime.InteropServices.Marshal]::SecureStringToBSTR($ss); $b64 = [System.Runtime.InteropServices.Marshal]::PtrToStringAuto($ptr); [System.Runtime.InteropServices.Marshal]::ZeroFreeBSTR($ptr); [System.Text.Encoding]::UTF8.GetString([Convert]::FromBase64String($b64))`;
    return execFileSync('powershell', ['-NoProfile', '-Command', ps], { encoding: 'utf8' }).trim();
  }
}

// --- Shared logic ---

function encrypt() {
  if (!fs.existsSync(ENV_PATH)) {
    console.error('.env file not found');
    process.exit(1);
  }
  const content = fs.readFileSync(ENV_PATH, 'utf8');
  const encrypted = platformEncrypt(content);
  fs.writeFileSync(SECURE_PATH, encrypted, { encoding: 'utf8', mode: 0o600 });
  const backend = IS_MACOS ? 'Keychain' : 'DPAPI';
  const user = process.env.USER || process.env.USERNAME;
  console.log(`Encrypted .env -> .env.encrypted (${backend}, tied to ${user})`);
  console.log('You can now delete .env if desired.');
}

function decrypt() {
  if (!fs.existsSync(SECURE_PATH)) return null;
  try {
    const encrypted = fs.readFileSync(SECURE_PATH, 'utf8').trim();
    return platformDecrypt(encrypted);
  } catch (e) {
    console.error('Failed to decrypt .env.encrypted:', e.message);
    return null;
  }
}

function loadSecureEnv() {
  const decrypted = decrypt();
  if (!decrypted) return false;
  for (const line of decrypted.split('\n')) {
    const trimmed = line.trim();
    if (!trimmed || trimmed.startsWith('#')) continue;
    const eqIdx = trimmed.indexOf('=');
    if (eqIdx < 0) continue;
    const key = trimmed.slice(0, eqIdx).trim();
    const value = trimmed.slice(eqIdx + 1).trim();
    if (!process.env[key]) {
      process.env[key] = value;
    }
  }
  return true;
}

if (require.main === module) {
  const cmd = process.argv[2];
  if (cmd === 'encrypt') {
    encrypt();
  } else if (cmd === 'decrypt') {
    const content = decrypt();
    if (content) {
      console.warn('[WARNING] Sensitive data follows:');
      console.log(content);
    } else {
      console.error('No .env.encrypted found or decryption failed');
    }
  } else {
    console.log('Usage: node secure-env.js <encrypt|decrypt>');
  }
}

module.exports = { loadSecureEnv };
